lego-powered-up/src/hubs/generic_hub.rs

```diff
@@ -0,0 +1,239 @@
+    //         ret.push(v.clone());
+    //     }
+
+    //     ret.sort_by_key(|x| x.port);
+
+    //     ret
+    // }
+
+    // Deprecated,  use enable_from_port or enable_from_kind
+    // async fn port(&self, port_id: Port) -> Result<Box<dyn Device>> {
+    //     let port =
+    //         *self.properties.port_map.get(&port_id).ok_or_else(|| {
+    //             crate::Error::NoneError(format!(
+    //                 "Port type `{port_id:?}` not supported"
+    //             ))
+    //         })?;
+    //     Ok(match port_id {
+    //         Port::HubLed => Box::new(devices::HubLED::new(
+    //             self.peripheral.clone(),
+    //             self.lpf_characteristic.clone(),
+    //             port,
+    //         )),
+    //         Port::A | Port::B  => {
+    //             Box::new(devices::RemoteButtons::new(
+    //                 self.peripheral.clone(),
+    //                 self.lpf_characteristic.clone(),
+    //                 port_id,
+    //                 port,
+    //             ))
+    //         }
+    //         _ => todo!(),
+    //     })
+    // }
+
+    /// Cache handles held by hub on device so we don't need to lock hub mutex as often    
+    fn device_cache(&self, mut d: IoDevice) -> IoDevice {
+        // Channels that forward some notification message types
+        d.cache_channels((
+            self.channels.singlevalue_sender.clone(),
+            self.channels.combinedvalue_sender.clone(),
+            self.channels.networkcmd_sender.clone(),
+        ));
+
+        // BT handles for calling send
+        d.cache_tokens((
+            Some(self.peripheral().clone()),
+            Some(self.characteristic().clone()),
+        ));
+
+        d
+    }
+
+    async fn io_from_port(&self, port_id: u8) -> Result<IoDevice> {
+        match self.connected_io.get(&port_id) {
+            Some(connected_device) => {
+                let mut d = connected_device.clone();
+                d = self.device_cache(d);
+
+                Ok(d)
+            }
+            None => Err(Error::HubError(format!(
+                "No device on port: {} ({:#x}) ",
+                &port_id, &port_id
+            ))),
+        }
+    }
+
+    async fn io_from_kind(&self, req_kind: IoTypeId) -> Result<IoDevice> {
+        let found: Vec<&IoDevice> = self
+            .connected_io
+            .values()
+            .filter(|&device| *device.kind() == req_kind)
+            .collect();
+        match found.len() {
+            0 => {
+                Err(Error::HubError(format!("No device of kind: {req_kind:?}")))   
+            }
+            1 =>  {
+                let device_deref = *found.first().unwrap();
+                let mut d = device_deref.clone();
+                d = self.device_cache(d);
+
+                Ok(d)
+            }
+            _ => {
+                Err(Error::HubError(format!("Found {:?} {req_kind:?} on {:?}, use io_from_port or io_multi_from_kind",
```

Comment:
```suggestion
                Err(Error::HubError(format!(
                concat!(
                    "Found {:?} {:?} on {:?}, ",
                    "use io_from_port or io_multi_from_kind"
                ),
                found.len(),
                req_kind,
                found.iter().map(|x| x.port()).collect::<Vec<_>>()
            )))
```
* move req_kind out of the format string for consistency with the other format args
* use concat! to split the format string & allow cargo fmt to shorten the line